charms/istio-pilot/src/resources_handler.py

```diff
@@ -0,0 +1,198 @@
+                elif "(Unauthorized)" in err.status.message and ignore_unauthorized:
+                    # Ignore error from https://bugs.launchpad.net/juju/+bug/1941655
+                    self.log.error(f"Ignoring unauthorized error:\n{err.status.message}")
+                else:
+                    self.log.error(err.status.message)
+                    raise
+            else:
+                raise
+
+    def delete_existing_resource_objects(
+        self,
+        resource,
+        namespace=None,
+        ignore_not_found=False,
+        ignore_unauthorized=False,
+        labels=None,
+    ):
+        if labels is None:
+            labels = {}
+
+        for obj in self.lightkube_client.list(
+            resource,
+            labels={"app.juju.is/created-by": f"{self.app_name}"}.update(labels),
+            namespace=namespace,
+        ):
+            self.delete_object(
+                obj,
+                namespace=namespace,
+                ignore_not_found=ignore_not_found,
+                ignore_unauthorized=ignore_unauthorized,
+            )
+
+    def apply_manifest(self, manifest, namespace=None):
+        for obj in codecs.load_all_yaml(manifest):
+            self.lightkube_client.apply(obj, namespace=namespace)
+
+    def delete_manifest(
+        self, manifest, namespace=None, ignore_not_found=False, ignore_unauthorized=False
+    ):
+        for obj in codecs.load_all_yaml(manifest):
+            self.delete_object(
+                obj,
+                namespace=namespace,
+                ignore_not_found=ignore_not_found,
+                ignore_unauthorized=ignore_unauthorized,
+            )
+
+    def create_ns_resource(self, filename: str, context: dict = None) -> GenericNamespacedResource:
+        """Returns a class representing a namespaced K8s resource.
+
+        Args:
+            - filename: name of the manifest file
+        """
+
+        # TODO: this is a generic context that is used for rendering
+        # the manifest files and extract their metadata. We should
+        # improve how we do this.
+        context = {
+            'namespace': 'namespace',
+            'app_name': 'name',
+            'name': 'generic_resource',
+            'request_headers': 'request_headers',
+            'response_headers': 'response_headers',
+            'port': 'port',
+            'service': 'service',
+        }
+
+        t = self.env.get_template(filename)
+        manifest = yaml.safe_load(t.render(context))
+        ns_resource = create_namespaced_resource(
+            group=manifest["apiVersion"].split("/")[0],
+            version=manifest["apiVersion"].split("/")[1],
+            kind=manifest["kind"],
+            plural=f"{manifest['kind']}s".lower(),
+            verbs=None,
+        )
+
+        return ns_resource
+
+    def reconcile_desired_resources(
+        self, resource: Any, desired_resources: Union[str, TextIO, None], namespace: str = None
```

Comment:
I'm not sure if the `namespace` arg works as expected.  The docstring mentions it is optional, but is it?  Does `lightkube_client.list(..., namespace=None)` work?  I'm expecting no, but maybe that just gets non-namespaced resources?  I'm also not sure if `.apply(..., namespace=None)` works or not (maybe it takes namespace from the Resource objects themselves, or it is for non-namespaced objects?).  In any case, if you're not sure either, maybe make namespace a required arg

Reply:
Left is as non-optional.